### PR TITLE
More explicit, and up to date docstrings for loaders.

### PR DESCRIPTION
**Pull request by gsaluja9**

### CODE REVIEW

File: .github/workflows/checks.yml [hunk 1]
```diff
@@ -14,19 +14,5 @@ jobs:
       with:
         python-version: '3.8.10'
     - uses: pre-commit/action@v2.0.3
-
     - uses: actions/checkout@v2
     - uses: luisremis/find-trailing-whitespace@master
-
```
Comment (luisremis) [inline .github/workflows/checks.yml:20]: why is this check going away? does PEP8 check already checks for trailing spaces?

File: aperturedb/BBoxLoader.py [hunk 2]
```diff
@@ -10,21 +10,22 @@
...
 
 class BBoxGeneratorCSV(CSVParser.CSVParser):
-    """**ApertureDB BBox Data loader.**
+    """**ApertureDB Bounding Box Data generator.**
 
     .. note::
-        Expects a csv file with the following columns:
+        Is backed by a csv file with the following columns:
 
             ``IMG_KEY``, ``x_pos``, ``y_pos``, ``width``, ``height``, ``BBOX_PROP_NAME_1``, ... ``BBOX_PROP_NAME_N``
```
Comment (luisremis) [inline aperturedb/BBoxLoader.py:18]: this (and any CSV generator, I believe) can also have "constraints_PROP" to specify conditional adding. Would you add that? And maybe an explanation in CSVGenerator about how that is used.

File: aperturedb/BlobLoader.py [hunk 3]
```diff
@@ -7,19 +7,29 @@
...
+        Is backed by a csv file with the following columns:
 
-            ``filename``, ``PROP_NAME_1``, ... ``PROP_NAME_N``, ``constraint_PROP1``
+            ``FILENAME``, ``PROP_NAME_1``, ... ``PROP_NAME_N``, ``constraint_PROP_NAME_1``
+
+    **FILENAME**: The path of the blob object on the file system.
+
+    **PROP_NAME_1 ... PROP_NAME_N**: Arbitrary property names associated with this blob.
+
+    **constraint_PROP_NAME_1**: Constraints against specific property, Used for uniqueness check.
```
Comment (luisremis) [inline aperturedb/BlobLoader.py:21]:
```suggestion
    **constraint_PROP_NAME_1**: Constraints against specific property, used for conditional adding the object.
```

File: aperturedb/ConnectionLoader.py [hunk 4]
```diff
@@ -7,19 +7,28 @@
...
-        has_image,321423532,42342522,0.4354,5432543254,5432543254
-        has_image,321423532,53241521,0.6432,98476542,98476542
+        ConnectionClass,_Image@id,_Descriptor@UUID,confidence,id,constaint_id
+        has_image,321423532,AID-0X3E,0.4354,5432543254,5432543254
+        has_image,42342522,BXY-AB1Z,0.6432,98476542,98476542
         ...
+
+    **ConnectionClass**: Arbitrary class name for the entity this would be saved as.
+
+    **<ClassName>@<PropertyName>** This is a special combination of Class Name and Property Name that can uniquely identify an entity.
```
Comment (luisremis) [inline aperturedb/ConnectionLoader.py:26]:
```suggestion
    **<ClassName>@<PropertyName>** This is a special combination of Class Name and 
    Property Name that can uniquely identify an entity.
```

Doesn't PEP8 check for line width?

File: aperturedb/ConnectionLoader.py [hunk 5]
```diff
@@ -76,14 +85,19 @@ def validate(self):
 class ConnectionLoader(ParallelLoader.ParallelLoader):
     """**ApertureDB Connection Loader.**
 
-        This class is to be used in combination with a "generator".
-        The generator must be an iterable object that generated "entity_data"
-        elements.
+    This executes in conjunction with a **generator** object for example :class:`~aperturedb.ConnectionLoader.ConnectionGeneratorCSV`,
+    which is a class that implements iterable inteface and generates "connection data" elements.
 
     Example::
```
Comment (luisremis) [inline aperturedb/ConnectionLoader.py:91]:
```suggestion
    Example:
```

File: aperturedb/ConnectionLoader.py [hunk 6]
```diff
@@ -76,14 +85,19 @@ def validate(self):
 class ConnectionLoader(ParallelLoader.ParallelLoader):
     """**ApertureDB Connection Loader.**
 
-        This class is to be used in combination with a "generator".
-        The generator must be an iterable object that generated "entity_data"
-        elements.
+    This executes in conjunction with a **generator** object for example :class:`~aperturedb.ConnectionLoader.ConnectionGeneratorCSV`,
```
Comment (luisremis) [inline aperturedb/ConnectionLoader.py:88]:
```suggestion
    This executes in conjunction with a **generator** object, 
    for example :class:`~aperturedb.ConnectionLoader.ConnectionGeneratorCSV`,
```

File: aperturedb/ConnectionLoader.py [hunk 7]
```diff
@@ -76,14 +85,19 @@ def validate(self):
...
+    which is a class that implements iterable inteface and generates "connection data" elements.
 
     Example::
 
-            entity_data = {
+            connection_data = {
                 "class":       connection_class,
+                "ref1_class": "_Image",
```
Comment (luisremis) [inline aperturedb/ConnectionLoader.py:95]: we should probably rename ref1 and ref2 as "src" and "dst", respectively, at some point to make it more explicit and consistent with our API

File: aperturedb/DescriptorLoader.py [hunk 8]
```diff
@@ -122,15 +143,15 @@ def validate(self):
...
+    which is a class that implements iterable inteface and generates "descriptor data" elements.
 
     Example::
 
-            image_data = {
-                "label":       label,
-                "properties":  properties,
+            descriptor_data = {
+                "label": label,
+                "set": set name,
```
Comment (luisremis) [inline aperturedb/DescriptorLoader.py:153]:
```suggestion
                "set": set_name,
```

File: aperturedb/DescriptorLoader.py [hunk 9]
```diff
@@ -122,15 +143,15 @@ def validate(self):
 class DescriptorLoader(ParallelLoader.ParallelLoader):
     """**ApertureDB Descriptor Loader.**
 
-    This class is to be used in combination with a "generator".
-    The generator must be an iterable object that generated "descriptor_data"
-    elements.
+    This class is to be used in combination with a **generator** object for example :class:`~aperturedb.DescriptorLoader.DescriptorGeneratorCSV`
```
Comment (luisremis) [inline aperturedb/DescriptorLoader.py:146]:
```suggestion
    This class is to be used in combination with a **generator** object, 
    for example :class:`~aperturedb.DescriptorLoader.DescriptorGeneratorCSV`
```

File: aperturedb/DescriptorLoader.py [hunk 10]
```diff
@@ -13,20 +13,41 @@
...
+        /mnt/data/embeddings/kitchen.npz,0,kitchen,kitchen_table,True,AID-0X3E,AID-0X3E
+        /mnt/data/embeddings/kitchen.npz,1,kitchen,kitchen_table,True,BXY-AB1Z,BXY-AB1Z
+        /mnt/data/embeddings/dining_chairs.npz,1,dining_chairs,special_chair,False,COO-SE1R,COO-SE1R
         ...
+
+    .. important::
+        In the above example, the index uniqely identiifes the actual np array from the many arrays in the npz file
+        which is same for line 1 and line 2. The UUID and constraint_UUID ensure that a Descriptor is inserted only once in the DB.
+
+        Association of an entity to a Descriptor can be specified by first ingesting some Entities, then Descriptors and finally by
```
Comment (luisremis) [inline aperturedb/DescriptorLoader.py:48]:
```suggestion
        Association of an entity to a Descriptor can be specified by first ingesting other Objects, then Descriptors and finally by
```

File: aperturedb/DescriptorSetLoader.py [hunk 11]
```diff
@@ -83,16 +83,16 @@ def validate(self):
 class DescriptorSetLoader(ParallelLoader.ParallelLoader):
     """**ApertureDB DescriptorSet Loader.**
 
-        This class is to be used in combination with a "generator".
-        The generator must be an iterable object that generated "entity_data"
-        elements.
+    This class is to be used in combination with a **generator** object for example :class:`~aperturedb.DescriptorSetLoader.DescriptorSetGeneratorCSV`
```
Comment (luisremis) [inline aperturedb/DescriptorSetLoader.py:86]:
```suggestion
    This class is to be used in combination with a **generator** object, 
    for example :class:`~aperturedb.DescriptorSetLoader.DescriptorSetGeneratorCSV`
```

File: aperturedb/ImageLoader.py [hunk 12]
```diff
@@ -193,9 +193,8 @@ class ImageLoader(ParallelLoader.ParallelLoader):
     """
     **ApertureDB Image Loader.**
 
-    This class is to be used in combination with a "generator".
-    The generator must be an iterable object that generated "image_data"
-    elements.
+    This class is to be used in combination with a **generator** object for example :class:`~aperturedb.ImageLoader.ImageGeneratorCSV`
```
Comment (luisremis) [inline aperturedb/ImageLoader.py:196]:
```suggestion
    This class is to be used in combination with a **generator** object, 
    for example :class:`~aperturedb.ImageLoader.ImageGeneratorCSV`
```

File: aperturedb/VideoLoader.py [hunk 13]
```diff
@@ -92,7 +92,7 @@ def validate(self):
 class VideoLoader(ParallelLoader.ParallelLoader):
     """**ApertureDB Video Loader.**
 
-    This class is to be used in combination with a :class:`~aperturedb.VideoLoader.VideoGeneratorCSV`.
+    This class is to be used in combination with a "generator", for example :class:`~aperturedb.VideoLoader.VideoGeneratorCSV`.
```
Comment (luisremis) [inline aperturedb/VideoLoader.py:95]:
```suggestion
    This class is to be used in combination with a "generator", 
    for example :class:`~aperturedb.VideoLoader.VideoGeneratorCSV`.
```

**Review by luisremis — COMMENTED** [review]
looks good. I added a couple of minor fixes.

File: .github/workflows/checks.yml (hunk 1 above)
Comment (gsaluja9) [reply .github/workflows/checks.yml:20]: I thought that pre-commit is something generic to run multiple github workflows. If you look at this PR, the trailing spaces seems to have run. Actually let's keep things simple for now, and let me reintroduce the previous configuration till I know of any benefits by doing it this way. Right now can't think of any apart from lesser code.

Comment (luisremis) [reply .github/workflows/checks.yml:20]: oh, never mind. I see what you are saying. this change is ok, we were actually running it twice before

Comment (drewaogle) [reply .github/workflows/checks.yml:20]: phew! that's what I thought was happening - but I've not used the actions a lot.

File: aperturedb/ConnectionLoader.py (hunk 4 above)
Comment (gsaluja9) [reply aperturedb/ConnectionLoader.py:26]: I don't think it runs on docstring (and thats good, or sphinx will get ruined)

File: aperturedb/ConnectionLoader.py (hunk 5 above)
Comment (gsaluja9) [reply aperturedb/ConnectionLoader.py:91]: That's a special directive for the auto api .

File: aperturedb/ConnectionLoader.py (hunk 7 above)
Comment (gsaluja9) [reply aperturedb/ConnectionLoader.py:95]: will do.

File: aperturedb/ConnectionLoader.py (hunk 5 above)
Comment (luisremis) [reply aperturedb/ConnectionLoader.py:93]: oh, my bad.

**Review by luisremis — APPROVED** [review]
looking great!

**Pull request merged**